### PR TITLE
Fix #20747 handle staff speed permanency/serialization/apply for new s…

### LINKED ISSUE #20747: Newly hired staff does not have the correct speed set when speed is not default (normal)
### Operating System

Ubuntu 20.04

### OpenRCT2 build

OpenRCT2, v0.4.5-271 (fa3339bdd on develop, DEBUG)

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

_No response_

### Describe the issue

When hiring new staff, that staff does not follow the speed set in the Cheats menu.

See https://github.com/OpenRCT2/OpenRCT2/pull/20734 for discussion.

Implement a proper cheet getter on the OpenRCT2/src/openrct2/Cheats.cpp level.

### Steps to reproduce

1. Open cheat menu
2. Navigate to "Park Cheats" tab
3. Modify the "Staff speed" option to "Fast" or "Frozen" (from its default state of "Normal")
4. Close the cheat menu
5. Hire a new member of staff
6. The new member of staff behaves as if "Staff speed" was set to "Normal" instead of what was set

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by KawkMob**
Staff speed is now permanent through:
* Cheat window open/close
* Game save/load
* Is applied to newly hired staff

This PR deprecates #20734

### CODE REVIEW

File: src/openrct2/Cheats.cpp [hunk 1]
```diff
@@ -235,6 +240,10 @@ void CheatsSerialise(DataSerialiser& ds)
                 case CheatType::MakeDestructible:
                     ds << gCheatsMakeAllDestructible;
                     break;
+                case CheatType::SetStaffSpeed:
+                LOG_ERROR("Deserializing staff speed at %d\n", gCheatsSelectedStaffSpeed);
```
Comment (ZehMatt) [inline src/openrct2/Cheats.cpp:244]: Not an error also shouldn't be here.

File: src/openrct2/park/ParkFile.cpp [hunk 2]
```diff
@@ -624,6 +624,7 @@ namespace OpenRCT2
         {
             os.ReadWriteChunk(ParkFileChunkType::CHEATS, [](OrcaStream::ChunkStream& cs) {
                 DataSerialiser ds(cs.GetMode() == OrcaStream::Mode::WRITING, cs.GetStream());
+                LOG_ERROR("ReadWriteCheatsChunk is saving %d\n", cs.GetMode() == OrcaStream::Mode::WRITING, cs.GetStream());
```
Comment (ZehMatt) [inline src/openrct2/park/ParkFile.cpp:627]: Remove this

**Review by ZehMatt — CHANGES_REQUESTED** [review]

**Comment by KawkMob** [comment]
PR cleaned and ready for re-review @ZehMatt

**Comment by ZehMatt** [comment]
Changes look good, you need to raise the save file format version and the network version and also this requires a changelog entry.

**Comment by KawkMob** [comment]
> Changes look good, you need to raise the save file format version and the network version and also this requires a changelog entry.

Some new stuff to explore - will look into this ;)

Any hints are much appreciated!

**Comment by KawkMob** [comment]
Updated - back to you, @ZehMatt

File: distribution/changelog.txt [hunk 3]
```diff
@@ -1,6 +1,7 @@
 0.4.7 (in development)
 ------------------------------------------------------------------------
 - Fix: [#20737] Spent money in player window underflows when getting refunds.
+- Fix: [#20747] Handle staff speed permanency/serialization/apply for new staff.
```
Comment (ZehMatt) [inline distribution/changelog.txt:4]:
```suggestion
- Fix: [#20747] Staff speed cheat not applying to newly hired staff, UI not showing the current applied speed.
```
The changelog should say what it fixes.

**Review by ZehMatt — CHANGES_REQUESTED** [review]
I think it looks good, left a remark on the changelog. This also needs to rebased.

File: distribution/changelog.txt (hunk 3 above)
Comment (KawkMob) [reply distribution/changelog.txt:4]: I'm currently on holidays oversees with no laptop. Will apply the fixes in approx 2weeeks.

**Comment by KawkMob** [comment]
Change applied, code rebased - back to you @ZehMatt

**Comment by ZehMatt** [comment]
Was flying over the PR again, I think it would be better if the new cheat variable has enum class type which has default/none as a member making this a bit more clear to what the variable is about.

**Comment by KawkMob** [comment]
> Was flying over the PR again, I think it would be better if the new cheat variable has enum class type which has default/none as a member making this a bit more clear to what the variable is about.

Just to double check. You'd like for gCheatsSelectedStaffSpeed to use something more user friendly instead of an index. 
Like for example the existing CHEATS_STAFF_FREEZE_SPEED - correct?

**Comment by ZehMatt** [comment]
> > Was flying over the PR again, I think it would be better if the new cheat variable has enum class type which has default/none as a member making this a bit more clear to what the variable is about.
> 
> Just to double check. You'd like for gCheatsSelectedStaffSpeed to use something more user friendly instead of an index. Like for example the existing CHEATS_STAFF_FREEZE_SPEED - correct?

Yup. The int32_t doesn't say much about it.

**Comment by KawkMob** [comment]
> > > Was flying over the PR again, I think it would be better if the new cheat variable has enum class type which has default/none as a member making this a bit more clear to what the variable is about.
> > 
> > 
> > Just to double check. You'd like for gCheatsSelectedStaffSpeed to use something more user friendly instead of an index. Like for example the existing CHEATS_STAFF_FREEZE_SPEED - correct?
> 
> Yup. The int32_t doesn't say much about it. However do not use the speed constants, they are a different thing.

Done. 
Have to say, was tempted just to use the speed constants just to avoid creating a new entity.

File: src/openrct2/Cheats.h [hunk 4]
```diff
@@ -11,6 +11,13 @@
 
 #include "common.h"
 
+typedef enum _WindowCheatsStaffSpeed
```
Comment (ZehMatt) [inline src/openrct2/Cheats.h:14]: Use enum class, also this has little to do with the window itsself. Use a name like `StaffSpeedCheat`

File: src/openrct2/Cheats.h [hunk 5]
```diff
@@ -11,6 +11,13 @@
...
+{
+    WINDOW_CHEATS_STAFF_SPEED_FROZEN,
```
Comment (ZehMatt) [inline src/openrct2/Cheats.h:16]: The first enum member should be None which means there is no cheat active.

File: src/openrct2/actions/StaffHireNewAction.cpp [hunk 6]
```diff
@@ -201,10 +201,25 @@ GameActions::Result StaffHireNewAction::QueryExecute(bool execute) const
...
+            case 1:
+                newPeep->Energy = CHEATS_STAFF_NORMAL_SPEED;
+                newPeep->EnergyTarget = CHEATS_STAFF_NORMAL_SPEED;
+                break;
+
+            default:
+                newPeep->Energy = CHEATS_STAFF_FAST_SPEED;
+                newPeep->EnergyTarget = CHEATS_STAFF_FAST_SPEED;
+                break;
+        }
```
Comment (ZehMatt) [inline src/openrct2/actions/StaffHireNewAction.cpp:220]: This needs to be re-worked. When the cheat is in-active it should just have the normal speed and only then should it check what the cheat overrides.

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2/actions/StaffHireNewAction.cpp (hunk 6 above)
Comment (KawkMob) [reply src/openrct2/actions/StaffHireNewAction.cpp]: Not entirely sure if I understand this correctly.

if gCheatsSelectedStaffSpeed is set to NONE, I still need to change energy to CHEATS_STAFF_NORMAL_SPEED in order for the peeps to go back to normal.

Hopefully the current implementation at least makes it clear what is going on.

**Comment by KawkMob** [comment]
Should be ready for review @ZehMatt  - one issue not set as resolved, please see comment.

One more fix that I'm tempted to do is to change the order of the available speeds in the game menu. Currently frozen is on top of the list. Would like for it to be the same as in the new cheats enum.

File: src/openrct2/actions/StaffHireNewAction.cpp (hunk 6 above)
Comment (ZehMatt) [reply src/openrct2/actions/StaffHireNewAction.cpp]: The default speed should be ordinarily set to normal, cheats override the value but also works for me when its set to default with `None`

File: src/openrct2-ui/windows/Cheats.cpp [hunk 7]
```diff
@@ -934,18 +958,24 @@ class CheatsWindow final : public Window
...
                     speed = CHEATS_STAFF_FREEZE_SPEED;
                     break;
+
                 case 1:
+                    gCheatsSelectedStaffSpeed = StaffSpeedCheat::NONE;
                     speed = CHEATS_STAFF_NORMAL_SPEED;
-            }
+                    break;
 
+                default:
```
Comment (ZehMatt) [inline src/openrct2-ui/windows/Cheats.cpp:974]: Don't use default for enum class types, handle each case explicitly. The compiler will emit a warning for when a case is not handled then.

File: src/openrct2/actions/StaffHireNewAction.cpp [hunk 8]
```diff
@@ -201,10 +201,25 @@ GameActions::Result StaffHireNewAction::QueryExecute(bool execute) const
...
+                newPeep->Energy = CHEATS_STAFF_FREEZE_SPEED;
+                newPeep->EnergyTarget = CHEATS_STAFF_FREEZE_SPEED;
+                break;
+
+            case StaffSpeedCheat::FAST:
+                newPeep->Energy = CHEATS_STAFF_FAST_SPEED;
+                newPeep->EnergyTarget = CHEATS_STAFF_FAST_SPEED;
+                break;
+
+            default:
```
Comment (ZehMatt) [inline src/openrct2/actions/StaffHireNewAction.cpp:216]: Also here, don't use default.

**Review by ZehMatt — CHANGES_REQUESTED** [review]
I think it's almost good. I think changing the order in the UI is also acceptable when it simplifies the code a bit.

**Comment by KawkMob** [comment]
Assuming the checks (hopefully) pass - back to you @ZehMatt

**Comment by ZehMatt** [comment]
> Assuming the checks (hopefully) pass - back to you @ZehMatt

Unfortunately they don't 😄 , please address the complaints. I think it looks reasonable now, after the final edits I can approve it.

**Comment by KawkMob** [comment]
> > Assuming the checks (hopefully) pass - back to you @ZehMatt
> 
> Unfortunately they don't 😄 , please address the complaints. I think it looks reasonable now, after the final edits I can approve it.

This is what I get for a late night push...

Anyway fixed, might have looked nicer with a 'default'  but this way it's at least readable.

Looking forward to this being pulled in!

**Comment by ZehMatt** [comment]
Looks good, I'll do some testing but I think its probably all fine.

File: distribution/changelog.txt [hunk 9]
```diff
@@ -14,6 +14,7 @@
 - Fix: [#20737] Spent money in player window underflows when getting refunds.
 - Fix: [#20778] [Plugin] Incorrect target api when executing custom actions.
 - Fix: [#20807] Tertiary colour not copied with small scenery.
+- Fix: [#20747] Staff speed cheat not applying to newly hired staff, UI not showing the current applied speed.
```
[the remark this answers is not in the document]
Comment (KawkMob) [reply distribution/changelog.txt]: Looks like it took me more than most to finish the bug ;)

Updated and in-test.

**Review by ZehMatt — APPROVED** [review]

**Comment by KawkMob** [comment]
Thanks for the approval @ZehMatt !

Anything else needed from my side before the merge?

File: src/openrct2-ui/windows/Cheats.cpp [hunk 10]
```diff
@@ -509,7 +508,7 @@ class CheatsWindow final : public Window
         // Current weather
         window_cheats_misc_widgets[WIDX_WEATHER].text = WeatherTypes[EnumValue(gClimateCurrent.Weather)];
         // Staff speed
-        window_cheats_misc_widgets[WIDX_STAFF_SPEED].text = _staffSpeedNames[_selectedStaffSpeed];
+        window_cheats_misc_widgets[WIDX_STAFF_SPEED].text = _staffSpeedNames[static_cast<int>(gCheatsSelectedStaffSpeed)];
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Cheats.cpp:511]:
```suggestion
        window_cheats_misc_widgets[WIDX_STAFF_SPEED].text = _staffSpeedNames[ToUnderlying(gCheatsSelectedStaffSpeed)];
```

If `gCheatsSelectedStaffSpeed` is an enum, we have this convenience function to get the int.

File: src/openrct2-ui/windows/Cheats.cpp [hunk 11]
```diff
@@ -851,7 +850,7 @@ class CheatsWindow final : public Window
                 WindowDropdownShowTextCustomWidth(
                     { windowPos.x + dropdownWidget->left, windowPos.y + dropdownWidget->top }, dropdownWidget->height() + 1,
                     colours[1], 0, Dropdown::Flag::StayOpen, 3, dropdownWidget->width() - 3);
-                Dropdown::SetChecked(_selectedStaffSpeed, true);
+                Dropdown::SetChecked(static_cast<int>(gCheatsSelectedStaffSpeed), true);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Cheats.cpp:853]: Same comment applies here.

File: src/openrct2-ui/windows/Cheats.cpp (hunk 10 above)
Comment (KawkMob) [reply src/openrct2-ui/windows/Cheats.cpp:511]: @tupaschoal - please elaborate. 

Do you mean std::to_underlying?
The other candidate could be ToUnderlying() from Identifier.hpp that's for class TIdentifier and returns a static cast anyway.

Comment (ZehMatt) [reply src/openrct2-ui/windows/Cheats.cpp:511]: We have a function called EnumValue which is doing the same as std::to_underlying, I think he mixed it up. It is defined in util/Util.h

Comment (tupaschoal) [reply src/openrct2-ui/windows/Cheats.cpp:511]: Whoops, @ZehMatt is right, sorry for the confusion!

Comment (KawkMob) [reply src/openrct2-ui/windows/Cheats.cpp]: > We have a function called EnumValue which is doing the same as std::to_underlying, I think he mixed it up. It is defined in util/Util.h

Thanks, done!

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**